For the treefrog converter

manodeep opened the pull request:
The idea is to create scripts that allow faster access for [sage](https://github.com/darrencroton/sage) but might benefit all semi-analytic codes

Commit:
Added gitignore

Commit:
Started adding MPI support.

Each rank will get a unique list of forestIDs to process and then write
to its own unique file.  The function `determine_forests()` is almost
complete but we simply need to do a broadcast (or perhaps just individual
send/recv).

Commit:
Treefrog_to_lhalo tentatively finished (including #1).

The function is MPI compatible and will generate multiple output files.

The function has been tested using the first 100 trees and successfully
runs on SAGE.

Commit:
Fixed populate forests function.

Commit:
Updated readme for using MPI. [ci skip]

Commit:
Fixed determine forests and added timers.

Commit:
Fixing up print statements.

Commit:
Added variable control how many Trees should be processed.

jacobseiler commented:
`SAGE` runs fine on the first 10,000 trees using this converter.  This is producing a reasonable SMF (given it's only a subset of the entire volume) so a good sign that everything is working as intended!  

Full tree generation is waiting in Slurm queue.  After the final trees are generated and `SAGE` SMF is confirmed I think we can pull this in.

manodeep commented:
Plots or it didn't happen :)

jacobseiler commented:
![1 stellarmassfunction](https://user-images.githubusercontent.com/17869663/43183185-95c67834-9027-11e8-96de-1cb9cec2087c.png)

manodeep commented:
Wait, what ? That doesn't look anything like the real Universe

jacobseiler commented:
It's only a fraction of the full volume so I'm not surprised it doesn't. I haven't applied any scaling to account for this.

jacobseiler commented:
I should've known you were correct!  The full trees converted but the results are completely messed up.  We forgot to properly calculate the Spin; currently it's using `Lx` `Ly` and `Lz`.

manodeep commented:
We also have a bunch of conversions to do. There is no little-h in any of the inputs, but we need to convert them out - perhaps check out how meraxes is doing it? Relevant code should be here: src/core/read_halos-velociraptor.c#L165-L180 (not posting link since meraxes is a private repo)

Commit:
Updated converter to do unit changes.

This involved putting little h into the mass and changing the spin parameter to
be normalized by the virial mass.

jacobseiler commented:
![1 stellarmassfunction](https://user-images.githubusercontent.com/17869663/43363657-45b16952-934d-11e8-9751-255b021d45e6.png)

![4 baryonictullyfisher](https://user-images.githubusercontent.com/17869663/43363612-f50acf94-934b-11e8-9fe4-8d77fb403352.png)
![8 blackholebulgerelationship](https://user-images.githubusercontent.com/17869663/43363613-f6fe05aa-934b-11e8-810f-0814b2676991.png)

Full trees with proper units conversions looks much better.  I've done a bit of parameter tinkering but haven't explored fully. I think with more we could fit the SMF etc much better...

manodeep commented on genesis/utils/treefrog_to_lhalo.py:
```diff
@@ -0,0 +1,738 @@
+                                         dtype=np.int32)
+
+        # Now loop over each tree and write the information to the HDF5 file.
+        for tree_idx in tqdm(range(NTrees)):
```
you can replace `tqdm(range(` with `trange(` with an associated `from tqdm import trange`

manodeep commented on genesis/utils/treefrog_to_lhalo.py:
```diff
@@ -0,0 +1,775 @@
+
+    # Update the `NextHaloInFOFgroup` chain the account for the new FoFs.
+    offset = 0
+    forest_halos = fix_nextsubhalo(forest_halos, [max_fof_mass_idx], offset, 
```
You don't need to call `fix_nextsubhalo`. Find the last subhalos of the `FOF` group (i.e., the subhalo which has a `-1` for the `NextSubhalo`) and then change that to point to this 'flyby' FOF.  I will acknowledge that changing it will make little difference to the total run time :)

manodeep commented on genesis/utils/treefrog_to_lhalo.py:
```diff
@@ -0,0 +1,775 @@
+    # If there is only one FoF Halo, no changes need to be made.
+    if len(root_halo_inds) == 1:
+        return forest_halos 
+
```
Check how many unique FoF halos there are.  If there's only one, you can also return immediately.

manodeep commented on genesis/utils/treefrog_to_lhalo.py:
```diff
@@ -0,0 +1,775 @@
+
+    # Since we're at the root snapshot, the indexing will start from 0.
+    root_halo_inds = np.arange(NHalos_root)
+    root_halos = forest_halos[root_halo_inds]
```
root_halos should probably point to only FOF halos (rather than all halos)

manodeep commented on genesis/utils/treefrog_to_lhalo.py:
```diff
@@ -0,0 +1,775 @@
+    global_flyby_ind = root_halo_inds[flyby_ind]
+
+    # Update them and flip the `MostBoundID` to flag the flyby.
+    forest_halos["FirstHaloInFOFgroup"][global_flyby_ind] = max_fof_mass_idx 
```
The RHS should be `root_halo_inds[max_fof_mass_idx]`

Commit:
Adjusted `fix_flybys()`.

The function now checks the FoF halos rather than all halos at the root snapshot.  Also added
a part following `fix_flybys()` that adjust `NextHaloInFOFgroup`.  Will need to add this
into a function.

jacobseiler commented:
@manodeep Can you please check the updated logic of `fix_flybys()`?  I updated it to consider only the main FoF halos.  In particular, can you confirm that the line `    forest_halos["FirstHaloInFOFgroup"][0:NHalos_root] = global_true_fof_idx` is correct? That is, if there is a flyby FoF with substructure (the subhalos point to the flyby FoF with `FirstHaloInFOFgroup`), then ALL substructure should be updated to point to the 'true_FoF'.  (Docstring hasn't been updated yet).

I moved the fixing of `NextHaloInFOFgroup` outside of it to keep it separate.  Can you please also check that logic is correct (starting from Line 387). 

I've run with 100 Trees and confirmed that `SAGE` runs on them (doesn't trip any of the assertion within `SAGE`).

manodeep commented:
Yup the substructure in the flyby FOF should now be pointing to the `global_tree_fof_idx`. 

I think you can fix the `NextSubhalo` for the entire tree in one shot, rather than one snapshot at a time. Or if you are assuming that subhalos are grouped with their FOF halo, then you should check that that is the case. Same for `fix_flybys`, if you are assuming a certain ordering of halos, make an assertion ensuring that. 

Otherwise, mostly looks good to me :) Great work!

jacobseiler commented:
The original idea for fixing `NextSubhalo` on the snapshot level was to avoid long search times with `np.where()`.  Since I'm only searching over those halos in the snapshot I wouldn't get slowdown for extraordinarily large trees.

But I agree that it's conceptually easier to grasp if I just pass the entire tree to `fix_nextsubhalo()` and have the entire tree be updated.  Perhaps if I looped over Snapshot within `fix_nextsubhalo()`?

manodeep commented:
If you impose the constraint of FOF+subhalos must be contiguous, then `NextSubhalo` is trivially `np.roll(np.arange(nsub, -1))` and then setting the last subhalo to have a -1

Commit:
Removed debugging prints.

Decided to keep the fixing of `NextSubhaloInFOFgroup` to be processed on a
snapshot by snapshot basis instead of the entire tree at once (see PR #2).

Commit:
Added explicit check that `NextHaloInFOFgroup` is contiguous.

jacobseiler commented:
I've decided to leave the fixing of `NextHaloInFOFgroup` on the snapshot level.  The reasoning is that doing it for the entire tree at once would require knowledge of the number of substructure for each main FoF halo (i.e., `Nsubs`).  While this IS present in the Genesis trees, I didn't want to impose this restriction on all trees.

This version has generated full trees for the `L500N1024` run and the SMF produced by SAGE (using fiducial inpute parameters) is attached.

![1 stellarmassfunction](https://user-images.githubusercontent.com/17869663/43692655-b6fd1af0-996c-11e8-9984-e092dc7afaef.png)

At this point there's just a few things left to do before we perform the pull

* Run Flake8 to ensure PEP8 compliance.
* Run Read-the-Docs.
* If I'm really feeling bored, implement a flag for writing out in HDF5 format (or both binary + HDF5).

Commit:
Merge branch 'treefrog_converter' of https://github.com/jacobseiler/genesis into treefrog_converter

Commit:
Updated to PEP8 compliance.

Commit:
Added flag to control whether to write a binary/HDF5 file (or both).

Need to adjust the writing of HDF5 to change the Position/Velocties/Spins
to Mx3 arrays (rather than individual x/y/z).

Commit:
Started working on way to turn Pos/Vel/Spin into Mx3 arrays.

(Temporary work so I can test on OzStar).

Commit:
Working version of turning Pos/Vel/Spin into Mx3.

Need to add Docs/cleanup.

Commit:
Conversion to multi-dimensional arrays working and properley documented.

Commit:
Working to get Sphinx running.

Currently issue is it's grabbing old docstrings for some reason.  Going to
reclone directory.

Commit:
Work to get Sphinx running.

Commit:
Got rid of Python3.4 checking (setup.py didn't require it).

Commit:
Added utilities README to index and update some docstrings.

Commit:
Added utilities file.

Commit:
Updated README to point to RTD [ci skip]
